stand/efi/loader/arch/amd64/ia32/elf64_freebsd32.c

```diff
@@ -0,0 +1,282 @@
+	NULL
+};
+
+#define PG_V	0x001
+#define PG_RW	0x002
+#define PG_PS	0x080
+
+#define GDT_P	0x00800000000000
+#define GDT_E	0x00080000000000
+#define GDT_S	0x00100000000000
+#define GDT_RW	0x00020000000000
+#define GDT_L	0x20000000000000
+
+#define M(x)	((x) * 1024 * 1024)
+
+typedef uint64_t p4_entry_t;
+typedef uint64_t p3_entry_t;
+typedef uint64_t p2_entry_t;
+typedef uint64_t gdt_t;
+
+static p4_entry_t *PT4;
+static p3_entry_t *PT3;
+static p3_entry_t *PT3_l, *PT3_u;
+static p2_entry_t *PT2;
+static p2_entry_t *PT2_l0, *PT2_l1, *PT2_l2, *PT2_l3, *PT2_u0, *PT2_u1;
+static gdt_t *GDT;
+
+extern EFI_PHYSICAL_ADDRESS staging;
+
+static void (*trampoline)(uint32_t stack, void *copy_finish, uint32_t kernend,
+    uint32_t modulep, uint64_t *pagetable, struct GDTR *gdtr, uint64_t entry);
+
+extern uintptr_t amd64_tramp;
+extern uint32_t amd64_tramp_size;
+
+/*
+ * There is an ELF kernel and one or more ELF modules loaded.
+ * We wish to start executing the kernel image, so make such
+ * preparations as are required, and do so.
+ */
+static int
+elf64_exec(struct preloaded_file *fp)
+{
+	struct file_metadata	*md;
+	struct GDTR	*gdtr;
+	Elf_Ehdr 		*ehdr;
+	vm_offset_t		modulep, kernend, trampcode, trampstack;
+	int			err, i;
+	bool			copy_auto;
+
+	/*
+	 * EFI_PHYSICAL_ADDRESS is 8 bytes wide on both 32 and 64-bit
+	 * systems, so we cant just use a pointer.
+	 */
+	EFI_PHYSICAL_ADDRESS ptr64;
+
+	copy_auto = copy_staging == COPY_STAGING_AUTO;
+	if (copy_auto)
+		copy_staging = fp->f_kernphys_relocatable ?
+		    COPY_STAGING_DISABLE : COPY_STAGING_ENABLE;
+
+	if ((md = file_findmetadata(fp, MODINFOMD_ELFHDR)) == NULL)
+		return (EFTYPE);
+	ehdr = (Elf_Ehdr *)&(md->md_data);
+
+	ptr64 = copy_staging == COPY_STAGING_ENABLE ?
+	    0x0000000040000000 /* 1G */ :
+	    0x0000000100000000; /* 4G */;
+	err = BS->AllocatePages(AllocateMaxAddress, EfiLoaderData, 1,
+	    &ptr64);
+	if (EFI_ERROR(err)) {
+		printf("Unable to allocate trampoline\n");
+		if (copy_auto)
+			copy_staging = COPY_STAGING_AUTO;
+		return (ENOMEM);
+	}
+	trampcode = ptr64;
+	bzero((void *)(uintptr_t)trampcode, EFI_PAGE_SIZE);
+	trampstack = trampcode + EFI_PAGE_SIZE - 8;
+	bcopy((void *)&amd64_tramp, (void *)(uintptr_t)trampcode,
+		amd64_tramp_size);
+	trampoline = (void *)(uintptr_t)trampcode;
+
+	ptr64 = copy_staging == COPY_STAGING_ENABLE ?
+	    0x0000000040000000 /* 1G */ :
+	    0x0000000100000000; /* 4G */;
+	err = BS->AllocatePages(AllocateMaxAddress, EfiLoaderData, 1,
+	    &ptr64);
+	if (EFI_ERROR(err)) {
+		printf("Unable to allocate GDT\n");
+		BS->FreePages(trampcode, 1);
+		if (copy_auto)
+			copy_staging = COPY_STAGING_AUTO;
+		return (ENOMEM);
+	}
+	GDT = (gdt_t *)(uintptr_t)ptr64;
+	bzero(GDT, EFI_PAGE_SIZE);
+	GDT[1] = GDT_P | GDT_E | GDT_S | GDT_RW | GDT_L; /* CS */
+	gdtr = (struct GDTR *)&GDT[2];
+	gdtr->ptr = (uintptr_t)GDT;
+	gdtr->size = 2 * sizeof(uint64_t) - 1;
+
+	if (copy_staging == COPY_STAGING_ENABLE) {
+		ptr64 = 0x0000000040000000; /* 1G */
```

Comment:
Picking a random spot... Maybe we should have constants for 1G and 4G...  I think existing code might be like that, though, and should be fixed too.